### PR TITLE
Remove redudant functions and code

### PR DESCRIPTION
**Pull request by Morlinest**
Remove useless functions/code. Fixes multiple bugs and remove duplicated code, like:

- Corrects ordering of repositories on admin page
- Corrects results when searching (explore page) - now you can get 2 different results when searching with/without keyword because two different functions are used and implementation is not same for both. But results with non empty keyword should always contains subset of full search (empty keyword) results.
- `models.Repositories()` ignores privacy settings in search but uses it for `count` (even it's always true as it is used only in 1 place with fixed privacy settings). This function can be replaced by `models.SearchRepositoryByName()` as it does exactly same thing.
- `models.GetRecentUpdatedRepositories()` doesn't make sense, as it is used only if query parameter `q` is empty and can return results with different ordering (not "recent updated"). `models.SearchRepositoryByName()` should do same thing.

Bonus: less code to maintain, less space for bugs :)

Extracted from #2371 and further improved.

### CODE REVIEW

File: routers/home.go [hunk 1]
```diff
@@ -155,9 +136,7 @@ func ExploreRepos(ctx *context.Context) {
 	ctx.Data["PageIsExploreRepositories"] = true
 
 	RenderRepoSearch(ctx, &RepoSearchOptions{
-		Ranger:   models.GetRecentUpdatedRepositories,
 		PageSize: setting.UI.ExplorePagingNum,
-		Searcher: ctx.User,
```
Comment (ethantkoenig) [inline routers/home.go]: Why is this line removed?

File: routers/home.go [hunk 2]
```diff
@@ -110,35 +109,17 @@ func RenderRepoSearch(ctx *context.Context, opts *RepoSearchOptions) {
...
-				return
-			}
-		}
+
+	repos, count, err = models.SearchRepositoryByName(&models.SearchRepoOptions{
+		Page:     page,
+		PageSize: opts.PageSize,
+		OrderBy:  orderBy,
+		Private:  opts.Private,
+		Keyword:  keyword,
```
Comment (ethantkoenig) [inline routers/home.go]: Should we add `Searcher: opts.Searcher,`?

**Review by ethantkoenig — CHANGES_REQUESTED** [review]
It seems that the `RepoSearchOptions.Searcher` field is never used. I think we still need to use it, but if not then we should just remove the field.

File: models/repo_list.go [hunk 3]
```diff
@@ -129,6 +129,8 @@ const (
 	SearchOrderByNewest                              = "created_unix DESC"
 	SearchOrderBySize                                = "size ASC"
 	SearchOrderBySizeReverse                         = "size DESC"
+	SearchOrderByID                                  = "ID ASC"
```
Comment (lafriks) [inline models/repo_list.go]: Please write column names lowercase

Comment (Morlinest) [reply models/repo_list.go]: Done

File: routers/home.go (hunk 2 above)
Comment (Morlinest) [reply routers/home.go]: `Searcher` is useless without `OwnerID`. In `models.SearchRepositoryByName` is used only if `OwnerID > 0`.

File: routers/home.go (hunk 1 above)
Comment (Morlinest) [reply routers/home.go]: Not needed anymore. ~But now I see I have to make some changes as `models.GetRecentUpdatedRepositories()` need only `Searcher` and not `OwnerID`.~

Comment (Morlinest) [reply routers/home.go]: @ethantkoenig I looked at it again and I don't need to change anything. Reason is to unify behaviour when you search with/without keyword. I'll post images.

**Comment by Morlinest** [comment]
@ethantkoenig This is how it looks now:

image
image

This PR should fix that. Explore page should be only for showing public repositories (as it is if you use any keyword). For private there is dashboard, user page, user settings, ...

**Comment by daviian** [comment]
@Morlinest Tested and seems good to me except one point: I miss owned private repositories in explore page (for none-admins).

**Comment by daviian** [comment]
@Morlinest Please rebase :)

**Comment by Morlinest** [comment]
@daviian Yes, it's expected change. The reason is explained in second and last point in PR description. Try to open explore page on current master build and type any keyword, you will be missing private repositories (example screenshots: https://github.com/go-gitea/gitea/pull/2652#issuecomment-334258612).

**Comment by daviian** [comment]
@Morlinest But instead of removing it for all results, it should be included IMO. But I think I am ok if we fix that in another PR.

**Comment by Morlinest** [comment]
I am not against it, but it would be few more changes to code. If also @lafriks or @ethantkoenig (or any other member :)) agree with it, I can change it to show private repositories (owned + collaborative) of signed user too.

**Comment by muellert** [comment]
I am very new to Gitea and have no Gogs experience, so my thinking may not be in line with the philosophy of the software at all, but I think item visibility should only depend on the role of the user performing a search and on the ACL for that item, not on which page the user started the search on.

**Comment by lafriks** [comment]
I also think that user should see on explore page all repositories he has access. For example in my company all repositories are private and in such case that would totally make that page useless

**Comment by Morlinest** [comment]
OK, I'll change current behaviour.

**Comment by muellert** [comment]
I wrote "items" because I don't know if repos are the only type of things affected. But I'd like to be able to find other things like issues or wiki pages as well, if I generally have access to them.

Where would I look in the code to figure all of this out?

**Review by ethantkoenig — APPROVED** [review]

File: models/repo_list.go [hunk 4]
```diff
@@ -182,6 +185,10 @@ func SearchRepositoryByName(opts *SearchRepoOptions) (repos RepositoryList, coun
 		cond = cond.And(builder.Eq{"is_private": false})
 	}
 
+	if opts.OwnerID > 0 && opts.AllPublic {
+		cond = cond.Or(builder.Eq{"is_private": false})
```
Comment (ethantkoenig) [inline models/repo_list.go:184]: I think this is a bug. If `opts.OwnerID > 0 && opts.AllPublic` is true, then we will search for all public repos, even those that do not match the keyword.

**Review by ethantkoenig — CHANGES_REQUESTED** [review]

File: models/repo_list.go (hunk 4 above)
Comment (Morlinest) [reply models/repo_list.go:184]: Fixed

**Review by ethantkoenig — APPROVED** [review]

**Comment by ethantkoenig** [comment]
LGTM

**Comment by Morlinest** [comment]
Added unit tests.

**Comment by daviian** [comment]
LGTM
